File: Silver II/BOJ_1260/BOJ_1260.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

int n, m, v;
vector<vector<int>> adj;

void DFS(int curNode, vector<bool>& isVisited)
{
    cout << curNode << ' ';
    isVisited[curNode] = true;

    int size = adj[curNode].size();
    for (int i = 0; i < size; ++i) {
        int nextNode = adj[curNode][i];
        if (!isVisited[nextNode]) {
            DFS(nextNode, isVisited);
        }
    }
}

void BFS(int startNode)
{
    vector<bool> isVisited(n + 1, false);
    queue<int> q;

    q.push(startNode);
    isVisited[startNode] = true;

    while (!q.empty()) {
        int curNode = q.front();
        q.pop();

        cout << curNode << ' ';

        int size = adj[curNode].size();
        for (int i = 0; i < size; ++i) {
            int nextNode = adj[curNode][i];

            if (!isVisited[nextNode]) {
                isVisited[nextNode] = true;
                q.push(nextNode);
            }
        }
    }
    cout << '\n';
}

int main()
{
    cin >> n >> m >> v;
    adj.resize(n + 1);
    for (int i = 0; i < m; ++i) {
        int v1, v2;
        cin >> v1 >> v2;

        adj[v1].push_back(v2);
        adj[v2].push_back(v1);
    }
    for (int i = 1; i <= n; ++i) {
        sort(adj[i].begin(), adj[i].end());
    }

    vector<bool> isVisited(n + 1, false);
    DFS(v, isVisited);
    cout << '\n';
    BFS(v);

    return 0;
}

File: Silver II/BOJ_1260/BOJ_1260.py
from collections import deque


def DFS(adj, start):
    visitedOrder = []
    stack = [start]

    while stack:
        cur = stack.pop()
        if cur not in visitedOrder:
            visitedOrder.append(cur)
            stack.extend(sorted(adj[cur], reverse=True))

    return visitedOrder


def BFS(adj, start):
    visitedOrder = []
    queue = deque()

    queue.append(start)
    while queue:
        cur = queue.popleft()  # Equeueual "queue.front(); queue.pop()" in C++
        if cur not in visitedOrder:
            visitedOrder.append(cur)
            queue.extend(sorted(adj[cur]))

    return visitedOrder


# Input
n, m, v = map(int, input().split())
adj = [set([]) for _ in range(n + 1)]
for _ in range(m):
    v1, v2 = map(int, input().split())
    adj[v1].add(v2)
    adj[v2].add(v1)

# Test
# Set은 자동정렬이 되나?
# for info in adj:
#     print(info)

# DFS & BFS
# 리스트에 *를 붙이면 '[' ']' ','를 제외하고 출력이 된다.
print(*DFS(adj, v))
print(*BFS(adj, v))
